### PR TITLE
Rework multiple fields to have selects

### PR DESCRIPTION
**Pull request by ClementPasteau**
Also update icon for all select in fields

image
image
Capture d’écran 2023-04-06 à 15 51 08
Capture d’écran 2023-04-07 à 18 25 45
Capture d’écran 2023-04-07 à 18 29 48
Capture d’écran 2023-04-07 à 18 29 55

### CODE REVIEW

**Comment by D8H** [comment]
These fields, the layer one especially, are often used in extensions and always as expression.
Can the field be an expression by default when extension events are edited?

**Comment by tristanbob** [comment]
Can you add then x and y positions to be displayed on the action sentence so it can be edited without needing to open the action popping?

File: newIDE/app/src/EventsSheet/ParameterFields/ExternalLayoutNameField.js [hunk 1]
```diff
@@ -7,35 +7,127 @@ import {
...
+
+    const switchFieldType = () => {
+      setIsExpressionField(!isExpressionField);
+    };
+
+    const onChangeSelectValue = (event, value) => {
+      props.onChange(event.target.value);
+    };
+
+    const onChangeTextValue = (value: string) => {
```
Comment (AlexandreSi) [inline newIDE/app/src/EventsSheet/ParameterFields/ExternalLayoutNameField.js:54]: Is it worth using callbacks for this method and the one above? I'm not sure

**Comment by AlexandreSi** [comment]
Will it break in app tutorials? I don't recall a tutorial when either of these fields are used

File: newIDE/app/src/EventsSheet/ParameterFields/ExternalLayoutNameField.js [hunk 2]
```diff
@@ -7,35 +7,127 @@ import {
...
+              label={<Trans>Select an external layout</Trans>}
+              onClick={switchFieldType}
+            />
+          ) : (
+            <RaisedButton
+              id="switch-expression-select"
+              icon={<Functions />}
+              style={style}
+              primary
+              label={<Trans>Use an Expression</Trans>}
```
Comment (AlexandreSi) [inline newIDE/app/src/EventsSheet/ParameterFields/ExternalLayoutNameField.js:126]: Should we display Expression capitalized but not external layout?

File: newIDE/app/src/EventsSheet/ParameterFields/ExternalLayoutNameField.js [hunk 3]
```diff
@@ -7,35 +7,127 @@ import {
...
     React.useImperativeHandle(ref, () => ({
       focus,
     }));
 
-    const externalLayoutNames: Array<ExpressionAutocompletion> = props.project
-      ? enumerateExternalLayouts(props.project).map(externalLayout => ({
-          kind: 'Text',
-          completion: `"${externalLayout.getName()}"`,
-        }))
+    const externalLayoutNames = props.project
```
Comment (AlexandreSi) [inline newIDE/app/src/EventsSheet/ParameterFields/ExternalLayoutNameField.js:32]: Use memo?

**Review by AlexandreSi — APPROVED** [review]
LGTM but I think it's worth adding what D8H said

**Comment by ClementPasteau** [comment]
> These fields, the layer one especially, are often used in extensions and always as expressions. Can the field be an expression by default when extension events are edited?

Tricky one as fields are abstracted and used without distinction in multiple places.
We could add some prop `preferExpressionIfAvailable` to the CommonParameters and pass it down depending on where it's used. But I'm afraid it would add a lot of prop drilling everywhere in the app

File: newIDE/app/src/EventsSheet/ParameterFields/ExternalLayoutNameField.js (hunk 3 above)
Comment (ClementPasteau) [reply newIDE/app/src/EventsSheet/ParameterFields/ExternalLayoutNameField.js:32]: probably an over optimization here imo

File: newIDE/app/src/EventsSheet/ParameterFields/ExternalLayoutNameField.js (hunk 1 above)
Comment (ClementPasteau) [reply newIDE/app/src/EventsSheet/ParameterFields/ExternalLayoutNameField.js:54]: probably yes, as they're passed down. I'll do that

**Comment by ClementPasteau** [comment]
> Can you add the x and y positions to be displayed on the action sentence so it can be edited without needing to open the action popping?

@tristanbob we could, and yet those props are optional, so I'm not sure what we do in the app regarding optional parameters usually. We don't show them?

**Comment by tristanbob** [comment]
> > Can you add the x and y positions to be displayed on the action sentence so it can be edited without needing to open the action popping?
> 
> @tristanbob we could, and yet those props are optional, so I'm not sure what we do in the app regarding optional parameters usually. We don't show them?

I feel like these position parameters are as "optional" as they are for the "create an object" action. (By this I mean, they should not be optional for either action)

By default, both actions will use (x=0, y=0) but having the position parameter visible in the sentence provides consistency and improves the speed for users.

**Comment by ClementPasteau** [comment]
> By default, both actions will use (x=0, y=0) but having the position parameter visible in the sentence provides consistency and improves the speed for users.

Completely agree with you on this one. I'll merge this one and do it as part of another PR as I'm afraid it'll not be as easy as it seems.
We need to find a way to display "0"/"0" when no value is entered, and not underlying the field in red because it's not filled.
I have an idea in mind

File: newIDE/app/src/EventsSheet/ParameterFields/ExternalLayoutNameField.js (hunk 2 above)
Comment (ClementPasteau) [reply newIDE/app/src/EventsSheet/ParameterFields/ExternalLayoutNameField.js:126]: I've aligned all namings with Capitalized letters

**Pull request merged**